custom_components/tapo/light.py

```diff
@@ -191,15 +204,17 @@ async def _change_brightness(self, new_brightness, apply_to_effect: str = None):
                 await self.device.set_brightness(new_brightness)
             ).get_or_raise()
 
-
+# follows https://developers.home-assistant.io/docs/core/entity/light/#color-modes
 def _components_to_color_modes(device: TapoBulb) -> set[ColorMode]:
-    color_modes = [ColorMode.ONOFF]
+    color_modes = []
     if device.is_color_temperature:
         color_modes.append(ColorMode.COLOR_TEMP)
-    if device.is_brightness:
-        color_modes.append(ColorMode.BRIGHTNESS)
     if device.is_color:
         color_modes.append(ColorMode.HS)
+    if device.is_brightness and not device.is_color_temperature and not device.is_color:
+        color_modes.append(ColorMode.BRIGHTNESS)
```

Comment:
I think you should just return here `{ColorMode.BRIGHTNESS}` and not append to the set, as the documentation states it should be the only possible color mode.
Same applies to `ONOFF`.

<img width="982" alt="Screenshot 2024-04-13 at 09 56 38" src="https://github.com/petretiandrea/home-assistant-tapo-p100/assets/2345875/65b7cc5c-2a6c-4456-b971-3c6e92553121">

So for example supported modes set to `color_temp, brightness` should never be possible.

Reply:
The code follows the HA algorithm describe here https://developers.home-assistant.io/docs/core/entity/light/#color-modes. Don't you agree?

Reply:
That's also the resource where my screenshot is from.
So this is not correct, as in the current implementation you have now there could potentially be scenarios where `brightness` could be mixed with other color modes in the `supported_color_modes`, but it has to be the only one.

See also the explanation how to correctly determine the color modes:

<img width="875" alt="Screenshot 2024-04-13 at 12 43 09" src="https://github.com/petretiandrea/home-assistant-tapo-p100/assets/2345875/ebb1eab3-3414-4f3c-8def-5792df28460b">

Reply:
The `if` has the `and not device.is_color_temperature and not device.is_color` so cannot be mixed with other color modes. No?

Reply:
Yes, I see you are correct about that.
However I see potential issue with the last condition:

```
if not device.is_color_temperature and not device.is_color:
        color_modes.append(ColorMode.ONOFF)
```

This does not check for `brightness`.
So you'll end up with `brightness, onoff` as supported color modes, which is not valid.

Maybe rewrite to:

```python
if not color_modes:
     color_modes.append(ColorMode.ONOFF)
```

Similarly as HA suggests in their approach.

Reply:
Ok, now It's clear and I agree with you